include/ginkgo/core/base/executor.hpp

```diff
@@ -1153,6 +1153,7 @@ class ExecutorBase : public Executor {
     void run(const Operation& op) const override
     {
         this->template log<log::Logger::operation_launched>(this, &op);
+        auto scope_guard = get_scoped_device_id_guard();
```

Comment:
very minor point: this adds another small allocation to run(...). It's unlikely this will affect performance in any way, but if it were, we could make the scoped guards non-polymorphic with an executor type enum.

Reply:
You are right, but I would do that in a separate PR.